circuit_database/plonky2/merkle_tree/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+        let expected_public_inputs = circuit_data.prover_only.public_inputs.clone();
+        for i in 0..4 {
+            pw.set_target(expected_public_inputs[i], tree.root.elements[i]);
+        }
+
+        // prove the circuit
+        let proof_with_pis = circuit_data.prove(pw).unwrap();
+
+        let verified = circuit_data.verify(proof_with_pis.clone()).unwrap();
+        println!("Verified: {:?}", verified);
+
+        Self {
+            proof: proof_with_pis,
+            verifier_only: circuit_data.verifier_only,
+            common: circuit_data.common,
+        }
+    }
+}
+
+pub fn verify_merkle_proof_circuit(
+    leaf_index: usize,
+    nr_layers: usize,
+) -> (
+    CircuitData<plonky2::field::goldilocks_field::GoldilocksField, PoseidonGoldilocksConfig, 2>,
+    Vec<HashOutTarget>,
+) {
+    let mut targets: Vec<plonky2::hash::hash_types::HashOutTarget> = Vec::new();
+
+    let config = CircuitConfig::standard_recursion_config();
+    let mut builder: CircuitBuilder<plonky2::field::goldilocks_field::GoldilocksField, 2> =
+        CircuitBuilder::<F, D>::new(config);
+
+    // The leaf to prove is in the Merkle Tree
+    let leaf_to_prove = builder.add_virtual_hash();
+    targets.push(leaf_to_prove);
+
+    // The first hashing outside of the loop, since it uses the leaf_to_prove
+    let merkle_proof_elm = builder.add_virtual_hash();
+    targets.push(merkle_proof_elm);
+
+    let mut next_hash: plonky2::hash::hash_types::HashOutTarget;
+    if leaf_index % 2 == 0 {
+        next_hash = builder.hash_or_noop::<PoseidonHash>(
+            [
+                leaf_to_prove.elements.to_vec(),
+                merkle_proof_elm.elements.to_vec(),
+            ]
+            .concat(),
+        );
+    } else {
+        next_hash = builder.hash_or_noop::<PoseidonHash>(
+            [
+                merkle_proof_elm.elements.to_vec(),
+                leaf_to_prove.elements.to_vec(),
+            ]
+            .concat(),
+        );
+    }
+
+    let mut current_layer_index = leaf_index / 2;
+
+    for _layer in 1..nr_layers {
+        let merkle_proof_elm = builder.add_virtual_hash();
+        targets.push(merkle_proof_elm);
+
+        if current_layer_index % 2 == 0 {
+            next_hash = builder.hash_or_noop::<PoseidonHash>(
+                [
+                    next_hash.elements.to_vec(),
+                    merkle_proof_elm.elements.to_vec(),
+                ]
+                .concat(),
+            );
+        } else {
+            next_hash = builder.hash_or_noop::<PoseidonHash>(
+                [
+                    merkle_proof_elm.elements.to_vec(),
+                    next_hash.elements.to_vec(),
+                ]
+                .concat(),
+            );
+        }
+        current_layer_index = current_layer_index / 2;
+    }
+    // This is the expected root value
+    builder.register_public_inputs(&next_hash.elements);
+
+    let data = builder.build::<C>();
+    (data, targets)
+}
```

Comment:
In general I'd like to see docstrings on the methods/functions in this file for clarity, especially since this is public-facing example code.

Reply:
Added comments to each method to  explain each step in the circuit construction:  [6b864c3](https://github.com/Sindri-Labs/sindri-resources/pull/98/commits/6b864c3fe1ebc81ea06bb683787db3ab7464a967)

Reply:
<img width="809" alt="Screenshot 2024-09-16 at 2 54 06 PM" src="https://github.com/user-attachments/assets/23d5fe1d-6639-4051-8840-87519cbae372">